### ISSUE TITLE
Discussion around a proper implementation of control-flow in pylint

### ISSUE THREAD

**Issue by Pierre-Sassoulas** [PyCQA/pylint#4795]
### Current problem

The main limitation regarding control flow is the fact that the variable checkers rely on ``NamesConsumer``'s to_consume/consumed states, and these are updated "all-or-nothing" per variable, with additional special-case checks here and there. To give a simpler example, pylint doesn't report any error on:

```python
def function():
    x = 10
    del x
    print(x)

function()
```

### Desired solution

I don't know the history of the Variables checker, but it seems to me like pylint will keep running into issues around control flow as long as this basic approach is used. I wouldn't say that this is a limitation of the AST visitor pattern though. I think improvements could be made by doing any of (in order of increasing code complexity and running time, but also correctness IMO):

    - Modifying ``NamesConsumer`` so that variables can be processed more than once.
    - Using astroid's variable lookup functionality, which handles control flow better. [1]
    - Building a control flow graph and then running ``VariablesChecker`` checks on the graph rather than the AST itself.

But all of these would be really significant changes so I'm not exactly advocating for them, just thought I'd mention. Personally I'm still very happy using pylint with the current functionality!

### Additional context

This issue is taken entirely from a comment by @david-yz-liu [here](https://github.com/PyCQA/pylint/pull/4791#issuecomment-892076763) and will serve as the reference for discussion regarding control-flow implementation.

**Comment by jacobtylerwalls** [comment]
Just noting for anyone following the discussion I've proposed baby steps in #5402 (and eventually #5384) following approach 1, above:

> Modifying NamesConsumer so that variables can be processed more than once.
